### PR TITLE
Requirement verification #182

### PR DESCRIPTION
**Pull request by lxatrhea**
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Visual appearence of Requirement verification process

Commit 1 (13-12):
- Update to latest pre-release CDP4-SDK version
- created ReactiveListMethods.DisposeAndClear and ReactiveListMethods.DisposeAndRemove for memory-leak-free disposing of objects in a ReactiveList<>
- Implemented IHaveContainerViewModel interface for looping through a visual tree (like in the Requirement browser).
- Created RequirementStateOfComplianceColorConverter for background color of rows in the RequirementBrowser that reflect the Requirement State Of Compliance for each individual row.
- Created RequirementStateOfComplianceToVisibilityConverter for showing and hiding the spinner when calculations/verification process is busy.
- Added more BooleanExpression (Extension) Methods and moved some to CDP4-SDK
- Created IHaveWritableRequirementStateOfCompliance interface to be used for visualisation reset after data (Thing) changes
- Added a button in the context menu that starts the verification process.
- Added ParametricConstraintsFolderRowViewModel for visual reasons (spinner) on the parametric constraints folder row in the Requirements Browser
- Setup subscriptions to events that change visuals concerning requirement verification
- Added spinner to requirement browser rows when verification/calculation is executed
- Added spinner control from GitHub (credits???)

Commit 2 (16-12):
- Added IsDeprecated filters

Commit 3 (17-12):
- CDP4-SDK upgrade
- Changed visuals (Requirement Browser and RelationshipBrowser -> BinaryRelationshipRowViewModel)
- Added Icon transparency
- 500 milliseconds delay before setting RequirementStateOfCompliance (Spinner visibility)

Commit 4 (17-12)
- Refactoring tooltip for RelationalExpression row in Requirementsbrowser
- Refactoring SetRequirementStateOfComplianceTree to update a visual tree in Requirements Browser when a (child) row was changed

Commit 5 (17-12):
- removed Commented text

### CODE REVIEW

File: Requirements/ViewModels/RequirementBrowser/Rows/RequirementRowViewModel.cs [hunk 1]
```diff
@@ -82,6 +91,19 @@ public class RequirementRowViewModel : CDP4CommonView.RequirementRowViewModel, I
...
+        /// <summary>
+        /// Gets or sets the <see cref="CDP4RequirementsVerification.RequirementStateOfCompliance"/>
+        /// </summary>
+        public RequirementStateOfCompliance RequirementStateOfCompliance
+        {
+            get => this.requirementStateOfCompliance;
+            set
+            {
+                this.RaiseAndSetIfChanged(ref this.requirementStateOfCompliance, value);
+                this.parametricConstraints.RequirementStateOfCompliance = value;
```
Comment (samatrhea) [inline Requirements/ViewModels/RequirementBrowser/Rows/RequirementRowViewModel.cs]: not so happy with this pattern, the reactive properties should just be that. Any logic that depends need to be implemented in a different manner. please listen to changes on this property in the constructor and have logic do work based on that.

File: Requirements/Converters/RequirementStateOfComplianceColorConverter.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+            if (Enum.TryParse<RequirementStateOfCompliance>(value.ToString(), out var relationalOperatorKind))
+            {
+                switch (relationalOperatorKind)
+                {
+                    case RequirementStateOfCompliance.Failed:
+                        color = "#FFACAC";
+
+                        break;
+
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply Requirements/Converters/RequirementStateOfComplianceColorConverter.cs]: Resharper did... Changed my local DotSettings file, which will become global in the near future.

File: Requirements/Extensions/BooleanExpressionExtensions.cs [hunk 3]
```diff
@@ -228,5 +188,76 @@ private static IReadOnlyList<BooleanExpression> GetMyExpressions(BooleanExpressi
 
             return myExpressions.ToList();
         }
+
+        /// <summary>
+        /// Gets the expressions that are direct children of <see cref="myself"/>
+        /// </summary>
+        /// <param name="myself"><see cref="BooleanExpression"/> for which the expressions should be found</param>
+        /// <returns><see cref="IReadOnlyList{BooleanExpression}"/> containing <see cref="BooleanExpression"/>s that are direct children of the class in the <see cref="myself"/> parameter</returns>
+        public static IReadOnlyList<BooleanExpression> GetAllMyExpressions(this BooleanExpression myself)
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply Requirements/Extensions/BooleanExpressionExtensions.cs]: The idea was that this (for now) would only be used for Requirement Verification and should therefore be part of the Requirement plugin.
I think we should also take a closer look if this is truly 10-25, or if it should remain our interpretation.
It's all about the interpretation of the BooleanExpression tree possibilities in Parametric Constraints.

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ParameterOrOverrideBaseRowViewModel.cs [hunk 4]
```diff
@@ -64,6 +64,11 @@ protected ParameterOrOverrideBaseRowViewModel(ParameterOrOverrideBase parameterO
             this.SetOwnerValue();
         }
 
+        /// <summary>
+        /// Gets or sets the <see cref="IThingCreator"/> that is used to create different <see cref="Things"/>.
+        /// </summary>
+        public IThingCreator ThingCreator { get; set; } = new ThingCreator();
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ParameterOrOverrideBaseRowViewModel.cs]: That means major refactoring is needed in (all!) RequirementBrowser and underlying ViewModels to make that possible.

File: Requirements/Extensions/BooleanExpressionExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (samatrhea) [reply Requirements/Extensions/BooleanExpressionExtensions.cs]: in the hand-coded POCO's there are many convenience methods like this, please move to CDP4Common.

**Pull request merged**